dask_expr/_expr.py

```diff
@@ -2239,6 +2240,40 @@ def are_co_aligned(*exprs):
 ## Utilites for Expr fusion
 
 
+def optimize_io_fusion(expr):
+    """Traverse the expression graph and apply fusion to the I/O layer that squashes
+    partitions together if possible."""
+
+    def _fusion_pass(expr):
+        new_operands = []
+        changed = False
+        for operand in expr.operands:
+            if isinstance(operand, Expr):
+                if (
+                    isinstance(operand, BlockwiseIO)
+                    and operand._fusion_compression_factor < 1
+                ):
+                    new = FusedIO(operand)
+                elif isinstance(operand, BlockwiseIO):
+                    new = operand
+                else:
+                    new = _fusion_pass(operand)
+
+                if new._name != operand._name:
+                    changed = True
+            else:
+                new = operand
+            new_operands.append(new)
+
+        if changed:
+            expr = type(expr)(*new_operands)
+
+        return expr
+
+    expr = _fusion_pass(expr)
+    return expr
```

Comment:
Maybe this could be implemented with the following:

```python
class BlockwiseIO:
    def simplify_down(self):
        if self._fusion_compression_factor < 1:
            return FusedIO(self)
```

I might be missing something here though.  

Maybe this is part of my confusion here.  This seems like an optimization that we always want to make on an expression type.  I think that that means that we don't really need to use the Expression stuff at all.  This could be a method on `IO`.

Sometimes people get used to using a fancy form of optimization that they forget about the simpler approaches they have.

Reply:
No sorry that won't work in simplify (see below). Happy to chat if my explanation is confusion. it has to come after combine_similar.